sdk/webpubsub/azure-webpubsub-client/azure/webpubsub/client/_client.py

```diff
@@ -0,0 +1,858 @@
+            elif parsed_message.kind == "disconnected":
+                handle_disconnected_message(parsed_message)
+            elif parsed_message.kind == "ack":
+                handle_ack_message(parsed_message)
+            elif parsed_message.kind == "groupData":
+                handle_group_data_message(parsed_message)
+            elif parsed_message.kind == "serverData":
+                handle_server_data_message(parsed_message)
+            else:
+                _LOGGER.warning("unknown message type: %s", parsed_message.kind)
+
+        def on_close(_: Any, close_status_code: int, close_msg: str):
+            if self._state == WebPubSubClientState.CONNECTED:
+                _LOGGER.info("WebSocket connection closed. Code: %s, Reason: %s", close_status_code, close_msg)
+
+                self._last_close_event = CloseEvent(close_status_code=close_status_code, close_reason=close_msg)
+                # clean ack cache
+                self._clear_ack_map()
+
+                if self._is_stopping:
+                    _LOGGER.warning("The client is stopping state. Stop recovery.")
+                    self._handle_connection_close_and_no_recovery()
+                    return
+
+                if self._last_close_event and self._last_close_event.close_status_code == 1008:
+                    _LOGGER.warning("The websocket close with status code 1008. Stop recovery.")
+                    self._handle_connection_close_and_no_recovery()
+                    return
+
+                if not self._protocol.is_reliable_sub_protocol:
+                    _LOGGER.warning("The protocol is not reliable, recovery is not applicable")
+                    self._handle_connection_close_and_no_recovery()
+                    return
+
+                recovery_url = self._build_recovery_url()
+                if not recovery_url:
+                    _LOGGER.warning("Connection id or reconnection token is not available")
+                    self._handle_connection_close_and_no_recovery()
+                    return
+
+                self._state = WebPubSubClientState.RECOVERING
+                recovery_start = time.time()
+                while (time.time() - recovery_start < _RECOVERY_TIMEOUT) and not self._is_stopping:
+                    try:
+                        self._connect(recovery_url)
+                        _LOGGER.debug("Recovery succeeded")
+                        return
+                    except:  # pylint: disable=bare-except
+                        _LOGGER.debug("Try to recover after %d seconds", _RECOVERY_RETRY_INTERVAL)
+                        time.sleep(_RECOVERY_RETRY_INTERVAL)
+
+                _LOGGER.warning("Recovery attempts failed after 30 seconds or the client is stopping")
+                self._handle_connection_close_and_no_recovery()
+            else:
+                _LOGGER.debug("WebSocket closed before open")
+                raise Exception(f"Fail to start Websocket: {close_status_code}")
```

Comment:
Same generic exception here - are customers going to see these errors? Looks like this should be some kind of IOError... Customers should never have to handle errors of type `Exception`

Reply:
https://github.com/Azure/azure-sdk-for-python/pull/28156#discussion_r1169780844